lib/vrl/parser/src/parser.lalrpop

```diff
@@ -0,0 +1,511 @@
+            .map(|(l, t, r)| (l, t.map(|s| s.to_owned()), r))
+            .collect();
+
+        let error = Error::ParseError {
+            span,
+            source,
+            dropped_tokens,
+        };
+
+        Node::new(span, RootExpr::Error(error))
+    },
+};
+
+// -----------------------------------------------------------------------------
+// expressions
+// -----------------------------------------------------------------------------
+
+Exprs: Vec<Node<Expr>> = {
+    Expr => vec![<>],
+    <v:(<Expr> EndOfExpression)+> <e:(<Expr>)?> => match e {
+        None => v,
+        Some(e) => {
+            let mut v = v;
+            v.push(e);
+            v
+        }
+    }
+};
+
+
+Expr: Node<Expr> = {
+    Sp<IfStatement> => Node::new(<>.span(), Expr::IfStatement(<>)),
+    AssignmentExpr,
+};
+
+EndOfExpression: () = {
+    "\n"+ => (),
+    ";" "\n"* => (),
+};
+
+NonterminalNewline: () = "\n";
+
+Ident: Ident = "identifier" => Ident(<>.to_owned());
+
+// An identifier that is allowed to include reserved keywords.
+AnyIdent: Ident = {
+    "identifier" => Ident(<>.to_owned()),
+    "reserved identifier" => Ident(<>.to_owned()),
+    "if" => Ident("if".to_owned()),
+    "else" => Ident("else".to_owned()),
+    "null" => Ident("null".to_owned()),
+    "true" => Ident("true".to_owned()),
+    "false" => Ident("false".to_owned()),
+};
+
+// -----------------------------------------------------------------------------
+// assignment
+// -----------------------------------------------------------------------------
+
+AssignmentExpr: Node<Expr> = {
+    Assignment => Node::new(<>.span(), Expr::Assignment(<>)),
+    ArithmeticExpr,
+};
+
+Assignment: Node<Assignment> = {
+    Sp<AssignmentSingle>,
+    Sp<AssignmentInfallible>,
+};
+
+AssignmentSingle: Assignment = {
+    <target: Sp<AssignmentTarget>> "="
+        <expr: Box<Expr>> => Assignment::Single { target, expr },
+}
+
+AssignmentInfallible: Assignment = {
+    <ok: Sp<AssignmentTarget>> ","
+        <err: Sp<AssignmentTarget>> "="
+        <expr: Box<Expr>> => Assignment::Infallible{ ok, err, expr},
+}
+
+#[inline]
+AssignmentTarget: AssignmentTarget = {
+    "_" => AssignmentTarget::Noop,
+    "." <Path?> => AssignmentTarget::External(<>),
+    Ident Path? => AssignmentTarget::Internal(<>),
+};
+
+// -----------------------------------------------------------------------------
+// arithmetic
+// -----------------------------------------------------------------------------
+
+ArithmeticExpr: Node<Expr> = {
+    Sp<ErrorCoalesce>,
+};
+
+ErrorCoalesce: Expr = {
+    Op<ErrorCoalesce, "??", Logical>,
+    Logical,
+};
+
+Logical: Expr = {
+    Op<Logical, "||", Equal>,
+    Op<Logical, "&&", Equal>,
+    Equal,
+};
+
+Equal: Expr = {
+    Op<Equal, "!=", Compare>,
+    Op<Equal, "==", Compare>,
+    Compare,
+};
+
+Compare: Expr = {
+    Op<Compare, ">=", Add>,
+    Op<Compare, ">", Add>,
+    Op<Compare, "<=", Add>,
+    Op<Compare, "<", Add>,
+    Add,
+};
```

Comment:
This should be:

```
Compare: Expr = {
    Op<Add, ">=", Add>,
    Op<Add, ">", Add>,
    Op<Add, "<=", Add>,
    Op<Add, "<", Add>,
    Add,
};
```

Since we don't allow expressions like `2 < 3 < 4`, we shouldn't recursively parse this. 

The same applies to `Equal` above. The following parses and actually allows this:

```
a = 3
b = 3
a == b == 3
```

But the above returns false, presumably because `a == b` evaluates to true, and `true != 3`.

Reply:
Right. So, I agree with your point in general, but not with the solution.

That is, the parser shouldn't derive any semantic meaning from the grammar, but the _compiler_ can (and I agree, should) reject these chained operations, and return a useful error message.

That is to say, the parser parses left to right, and similar to how `2 * 2 * 2` is interpreted by the compiler as meaning `(2 * 2) * 2`, that's the same how the compiler _currently_ interprets `2 > 2 > 2` (which thus gets turned into `false > 2`, which is a compile-time error).

So, my conclusion is that we should reject chaining comparison operators in the compiler, but not in the parser.

I'll fix this before the PR is merged.

I suspect we'll end up with an error message similar to Rust's:

```
error: comparison operators cannot be chained
 --> src/main.rs:2:22
  |
2 |     println!("{}", 2 == 3 == 4);
  |                      ^^   ^^
  |
help: split the comparison into two
  |
2 |     println!("{}", 2 == 3 && 3 == 4);
  |                           ^^^^
```

```
error: comparison operators cannot be chained
 --> src/main.rs:2:22
  |
2 |     println!("{}", 2 < 3 < 4);
  |                      ^   ^
  |
help: split the comparison into two
  |
2 |     println!("{}", 2 < 3 && 3 < 4);
  |                          ^^^^
```

Reply:
fixed locally